---
title: "Somatic Variant Calling"
author: "BIOF2014"
format:
  pdf:
    include-in-header:
      - text: |
          \usepackage{amsfonts}
          \usepackage{amsmath}
          \usepackage{amssymb}
          \usepackage{amsthm}
          \usepackage{fixmath}
          \usepackage[english]{babel}
          \usepackage{mathrsfs}
          \usepackage{mathtools}
---

## Motivation

Building on our modeling experience in germline variant calling, we will now
establish a model for somatic variant calling.

In cancer sequencing, we are interested in identifying variants that are 
in present in the cancer cells but not the normal cells. These are known
as **somatic variants** (or mutations).

Let $g_N$ represent the genotype in normal cells at a genomic position $l$,
and let $g_T$ represent the genotype in tumour cells. The joint genotypes are as follows.

|           $g_T$ = | AA        | AB       | BB       |
|------------------:|-----------|----------|----------|
| $g_N = \text{AA}$ | wild type | somatic  | somatic  |
| AB                | LOH       | germline | LOH      |
| BB                | error     | error    | germline |

: Joint genotype of a tumour-normal pair

LOH, loss of heterozygosity.

Since tumour cells come from normal cells, $g_T$ is dependent on $g_N$, and
some joint genotypes are likely to be due to sequencing errors. Therefore,
calling the variants in a tumour sample jointly with its matched normal
sample would yield higher accuracy than calling the genotype for normal and 
tumour samples separately.

Additionally, we will also enhance our previous variant calling model 
by accounting for whether a read is mapped correctly using its 
**mapping quality score**.


## Somatic variant model for a single tumour

Our model will assess each genomic position $l$ separately, so we
will omit the $l$ index.

The input consists of $J$ reads with observed read status 
$\mathbold{X} = [X_j]$, base correctness probabilities
$\mathbold{q} = [q_j]$, and mapping correctness probabilities 
$\mathbold{r} = [r_j]$.
$X_j = 1$ if read $j$ contains the alternative allele at genomic position
$l$, and $X_j = 0$ otherwise.
$q_j$ represents probability that the base mapping to genomic position 
$l$ is read correctly.
In other words, $q_j = 1 - e_j$ where $e_j$ is the probability of 
read error as defined in the previous tutorial.
$r_j$ represents the probability that read $j$ is mapped
correctly.
We will treat $x_j$ as observed random variables, while
$q_j$ and $r_j$ are observed fixed values.

We define $Z_j$ as the latent indicator random variable for 
$X_j$ such that $Z_j = 1$ if read $j$ truly contains the 
alternative allele and $Z_j = 0$ otherwise.

We further $C_j$ as the latent indicator random variable such that
$C_j = 1$ if read $j$ is mapped correctly and $C_j = 0$ otherwise.

If read $j$ is mapped correctly ($C_j = 1$), the likelihood is 
similar to the germline variant model:
$$
\begin{aligned}
p \left( x_j \mid Z_j = 1, C_j = 1 \right)
&= \begin{cases}
 q_j & \text{ if } x_j = 1 \\
 1 - q_j & \text{ if } x_j = 0
\end{cases} \\
p \left( x_j \mid Z_j = 0, C_j = 1 \right)
&= \begin{cases}
 1 - q_j & \text{ if } x_j = 1 \\
 q_j & \text{ if } x_j = 0
\end{cases} .
\end{aligned}
$$

If read $j$ is mapped incorrectly ($C_j = 0$), the likelihood is
uniform over the domain of $x_j$:
$$
p \left( x_j \mid Z_j = 1, C_j = 0 \right) = \frac{1}{2},
\quad x_j \in \{0, 1\} .
$$

The prior on the latent mapping indicator $C_j$ is given by
$$
p \left( c_j \right)
= \begin{cases}
 r_j & \text{ if } c_j = 1 \\
 1 - r_j & \text{ if } c_j = 0
\end{cases} .
$$

Now, the prior on the latent read indicator $Z_j$ is a bit different
from the germline variant model.
$$
p \left( z_j \mid g, \mu_g \right) = 
\begin{cases}
 \mu_g & \text{ if } z_j = 1 \\
 1 - \mu_g & \text{ if } z_j = 0
\end{cases} .
$$
In other words, $\mu_g$ is the probability that $z_j = 1$ given genotype 
$g \in \{0, 1, \ldots, M\}$, which represents the number of copies of 
the alternative allele.

Under the germline variant model, we derived a general expression
for $p(z_j \mid g, m)$ based on genotype $G$ and ploidy $M$. 
Recall that
$$
p(z_j \mid g, m) =
\begin{cases}
 \frac{g}{m} & \text{ if } z_j = 1 \\
 1 - \frac{g}{m} & \text{ if } z_j = 0
\end{cases}  .
$$
Therefore, $\mu_g = \frac{g}{m}$ for germline variants.

However, under a somatic variant model, $\mu_g$ is more difficult to
define because it now depends on tumour ploidy, purity, and copy-number
state at the genomic position. Rather than defining $\mu_g$ directly,
we can define it probabilitistic by putting a prior on $\mu_g$.
$$
p \left( \mu_g \right) = \text{Beta} \left( \mu_g \mid a_g,  \right) ,
$$
where $a_g$ and $b_g$ are model **hyperparameters** (i.e. tuning parameters
that we can set to specific values). Hyperparameters are normally treated
as fixed values.

Now, we are ready to define the prior on genotype $G$:
$$
p \left( G = g \mid \symbf{\theta} \right) 
= \text{Categorical}\left( g \mid \symbf{\theta} \right) ,
\quad g \in \{0, 1, \ldots, m\} ,
$$

In turn, we put a prior on the parameter $\symbf{\theta}$:
$$
p \left( \symbf{\theta} \right)
= \text{Dirichlet}\left( \symbf{\theta} \mid \mathbold{d} \right) ,
$$
where $\mathbold{d} = [ d_g ]$ are hyperparameter weights.

Hence, we have extended our germline variant model to consider mapping errors
and to incorporate uncertainty or variability in $\mu_g$, as well as putting
a flexible prior on the genotype $G$.
The model that we described so far closely resembles the [SNVMix2 model][1],
though we have fixed a few errors in the latter model.


## Somatic variant model for a tumour-normal sample pair

Previously, we defined the unknown genotype parameter $G$ as the
number of copies of the alternative allele in the genome of the sample
at position $l$. Now, we have a tumour-normal pair, and we need to
formally define $G$ to correspond to the joint genotype of tumour-normal 
pair (Table 1).

Therefore, define joint genotype set as
$$
\begin{aligned}
\mathcal{S} = \{
  &(AA,AA), (AB,AA), (BB,AA), \\
  &(AA,AB), (AB,AB), (BB,AB), \\
  &(AA,BB), (AB,BB), (BB,BB)
\} ,
\end{aligned}
$$
where we enumerate the joint genotype table in column-major order.
Then, random variable $G = k$ corresponds to the $k$th element of 
$\mathcal{S}$.

Finally, we can define $\mathbold{d}$ such that more probable
joint genotypes are given higher weights, and $d_k$ also
corresponds to the $k$th element of $\mathcal{S}$.

Hence, we have a model similar to the [JointSNVMix2 model][2].


## Posterior

For variant calling, we would want to solve for the posterior 
distribution $p \left( g \mid \mathbold{x} \right)$ for the 
random variable $G$ with support $\mathcal{G}$, which can be
derived using Bayes' theorem.
$$
p \left( g \mid \mathbold{x} \right)
= \frac{ p \left( \mathbold{x} \mid g \right) \, p \left( g \right) }
       { p \left( \mathbold{x} \right) } .
$$

The prior $p\left(g\right)$ is not immediately available from our model
definition, so we need to derive it by
$$
\begin{aligned}
p \left( G \right)
&= \int_{\symbf{\Theta}} \;
   p \left( G \mid \symbf{\theta} \right) \,
   p \left( \symbf{\theta} \right) \, d \symbf{\theta} \\
&= \int_{\symbf{\Theta}} \;
   \symbf{\theta} \;
   p \left( \symbf{\theta} \right) \, d \symbf{\theta} \\
&= \text{E} \left( \symbf{\theta} \right)  
\quad (\text{definition of expectation}) .
\end{aligned}
$$

Therefore, 
$$
p \left( g \right)
= \text{E} \left( \theta_g \right)
$$

Under the Dirichlet distribution, the expectation is given by
$$
\text{E} \left( \theta_g \right)
= \frac{d_g}{\sum_k d_g} .
$$

On the other hand, the model evidence $p \left( \mathbold{x} \right)$ is given by
$$
p \left( \mathbold{x} \right) 
= \sum_{g \in \mathcal{G}} 
  p \left( \mathbold{x} \mid g \right) \, p \left( g \right) .
$$

Now, we need to derive the following quantities:
$$
\begin{aligned}
p \left( \mathbold{x} \mid g \right) 
&= \sum_j p \left( x_j \mid g \right) \\
p \left( x_j \mid g \right)
&= \int_0^1 p \left( x_j \mid g, \mu_g \right) 
   p \left( \mu_g \right) \, d \mu_g \\
p \left( x_j \mid g, \mu_g \right)
&= \sum_{c=0}^1 \sum_{z=0}^1
  p\left(x_j, g, \mu_g, c, z \right) .
\end{aligned}
$$

<!--
p \left( x_j \mid g, \mu_g \right)
= \sum_{c=0}^1 \sum_{z=0}^1
  p\left(x_j \mid c, z \right) \, 
  p\left( c \right) \,
  p\left( z \mid g, \mu_g \right) .
-->

## Questions

1. Verify that each probability distribution that we define is proper
   (i.e. the distribution satisfies Kolmogorov's axiom 1 and 2).
   
2. Assign reasonable values to hyperparamter $\mathbold{d}$ in the
   somatic model for a single tumour.

3. Using the information in Table 1, assign appropriate values to 
   the hyperparameter $\mathbold{d}$ in the somatic model for
   tumour-normal pair.

4. What does Table 1 assume about the ploidy of the tumour?

5. If you were to explicitly show the genomic position index $l$,
   which variables would be indexed by $l$?
   
6. If you were to explicity show the sample index $i$,
   which variables would be indexed by $i$?


[1]: https://pmc.ncbi.nlm.nih.gov/articles/PMC2832826/pdf/btq040.pdf
[2]: https://pmc.ncbi.nlm.nih.gov/articles/PMC3315723/pdf/bts053.pdf
